### PR TITLE
Removes hard coded App Insights location

### PR DESCRIPTION
**Pull request by kyleburnsdev**
Allows Application Insights to be deployed to the same location as the other resources. Existing code appears to be from when App Insights was not widely available

# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [X] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Comment by mcollier** [comment]
If this template is deployed to a region that does not yet support Application Insights, will it fail?  

Alternative, could there be a separate parameter, such as `applicationInsightsLocation` (with a default value of `resourceGroup().location`)?  This would allow the user to override the region for Application Insights to ensure they can Application Insights to a supported region.

**Comment by kyleburnsdev** [comment]
> If this template is deployed to a region that does not yet support Application Insights, will it fail?
> 
> Alternative, could there be a separate parameter, such as `applicationInsightsLocation` (with a default value of `resourceGroup().location`)? This would allow the user to override the region for Application Insights to ensure they can Application Insights to a supported region.

@mcollier - reviewing the supported regions list, I think you're right. I'll update the PR

**Comment by mcollier** [comment]
LGTM

File: 101-function-premium-vnet-integration/azuredeploy.json [hunk 1]
```diff
@@ -6,7 +6,14 @@
             "type": "string",
             "defaultValue": "[resourceGroup().location]",
             "metadata": {
-                "description": "Location for all resources."
+                "description": "Location for all resources except Application Insights."
+            }
+        },
+        "appInsightsLocation": {
+            "type": "string",
+            "defaultValue": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline 101-function-premium-vnet-integration/azuredeploy.json:14]: AI is not globally available so you can't have a default that can resolve to any location.  You'll need to add param files for Public and Gov to set it.

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: 101-function-premium-vnet-integration/azuredeploy.json (hunk 1 above)
Comment (mcollier) [reply 101-function-premium-vnet-integration/azuredeploy.json:14]: Ah, so an azuredeploy.parameters.us.json file - https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#sample-application-files

And then not set a default value for `appInsightsLocation`.

Comment (kyleburnsdev) [reply 101-function-premium-vnet-integration/azuredeploy.json:14]: @bmoore-msft  - I see what you're getting at, but also am concerned about the usability of removing the default. Most regions currently have AI or will very soon, so there's a balance to be considered on placing the burden of specifying the value (and associated risk of having it be cross region when specified) against the risk that someone on an unsupported region will initially submit the deployment accepting the default and have to re-submit with an updated value.

Do you have thoughts on this balance?

Comment (bmoore-msft) [reply 101-function-premium-vnet-integration/azuredeploy.json:14]: a defaultValue should never cause a deployment to fail and it's only available in about half the locations, so...  we don't want to use something that's not very likely to work.

Comment (kyleburnsdev) [reply 101-function-premium-vnet-integration/azuredeploy.json]: @bmoore-msft I've updated to favor the current deployment region while falling back to the original behavior when that region is not known to have Application Insights available

Comment (bmoore-msft) [reply 101-function-premium-vnet-integration/azuredeploy.json]: Sorry but I think we've done a 180 here... the original template was simple and could not fail due to user input.  

Now we're asking not only to select a location for AI to supply a few other params including a json array of all the locations that are supported (which is not easy to find).  This is far more complex.

I can think of a few options here:
1) the original template
2) add a distinct param for AI location that the user must supply
3) Option 2 but make the param optional and use the approach in the original template to supply the value when not provided.

Comment (kyleburnsdev) [reply 101-function-premium-vnet-integration/azuredeploy.json]: Option 3 is pretty much the approach used here. The valid values are provided in the cloud-specific param file and the user does not have to supply. 

- If the user specifies a value for AI location, it is used. 
- If the user does not supply a value and the location is in the list of valid values, the location is used (reducing latency and egress exposure).
- If the user does not supply a value and the location is _not_ in the list of valid values, the values from the original template are used as a default

The only change that the user of the template experiences is the ability to optionally specify a location for AI without breaking any current scripts that may have been created

Comment (bmoore-msft) [reply 101-function-premium-vnet-integration/azuredeploy.json]: It's not really the same as what I'm suggesting - the approach in the current template has 2 additional required parameters that need to be supplied by the user and even then it's not very clear which location is actually going to be used... also, the parameter files is not part of the template...  This needs to be simplified if you want to merge the PR.

**Pull request merged**